src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -151,24 +159,43 @@ private void FlattenConcatArg(BoundExpression lowered, ArrayBuilder<BoundExpress
         /// its args if so.
         /// </summary>
         /// <returns>True if this is a call to a known string concat operator, false otherwise</returns>
-        private bool TryExtractStringConcatArgs(BoundExpression lowered, out ImmutableArray<BoundExpression> arguments)
+        private bool TryExtractStringConcatArgs(BoundExpression lowered, out ImmutableArray<BoundExpression> arguments, out ImmutableArray<LocalSymbol> previousLocals)
         {
-            switch (lowered.Kind)
+            previousLocals = default;
+
+            switch (lowered)
             {
-                case BoundKind.Call:
-                    var boundCall = (BoundCall)lowered;
+                case BoundCall boundCall:
                     var method = boundCall.Method;
                     if (method.IsStatic && method.ContainingType.SpecialType == SpecialType.System_String)
                     {
-                        if ((object)method == (object)_compilation.GetSpecialTypeMember(SpecialMember.System_String__ConcatStringString) ||
-                            (object)method == (object)_compilation.GetSpecialTypeMember(SpecialMember.System_String__ConcatStringStringString) ||
-                            (object)method == (object)_compilation.GetSpecialTypeMember(SpecialMember.System_String__ConcatStringStringStringString))
+                        if ((object)method == _compilation.GetSpecialTypeMember(SpecialMember.System_String__ConcatStringString) ||
+                            (object)method == _compilation.GetSpecialTypeMember(SpecialMember.System_String__ConcatStringStringString) ||
+                            (object)method == _compilation.GetSpecialTypeMember(SpecialMember.System_String__ConcatStringStringStringString))
+                        {
+                            arguments = boundCall.Arguments;
+                            return true;
+                        }
+
+                        // Faced `string.Concat(ReadOnlySpan<char>, ReadOnlySpan<char>) + string` (e.g. as a result of a previous rewrite).
+                        // We can only unwrap this if we can later rewrite it to `string.Concat(ReadOnlySpan<char>, ReadOnlySpan<char>, ReadOnlySpan<char>)`
+                        if ((object)method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpan) &&
+                            TryGetWellKnownTypeMember<MethodSymbol>(lowered.Syntax, WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpanReadOnlySpan, out _, isOptional: true))
                         {
                             arguments = boundCall.Arguments;
```

Comment:
>arguments = boundCall.Arguments; [](http://example.com/codeflow?start=28&length=32)

Can we assert that ```WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar``` and ```WellKnownMember.System_ReadOnlySpan_T__ctor_Reference``` if we get here? If we can, it would be good to do so with a comment explaining why.